python/ray/tune/search/placeholder.py

```diff
@@ -0,0 +1,124 @@
+from typing import Any, Dict, Tuple
+from ray.tune.search.sample import Categorical, Function
+from ray.tune.search.variant_generator import assign_value, _get_value
+
+from ray.util.annotations import DeveloperAPI
+
+
+class _Categorical_Value:
+    """Replaced value for categorical typed objects.
+    """
+    def __init__(self, choices):
+        self._choices = choices
+
+    def get(self, ph):
+        ca, i = ph.split("_")
+        assert ca == "cat", "Categorical placeholder should start with cat_"
+        return self._choices[int(i)]
+
+    def get_ph(self):
+        return [f"cat_{i}" for i in range(len(self._choices))]
+
+
+class _Function_Value:
+    """Replaced value for function typed objects.
+    """
+    def __init__(self, fn):
+        self._fn = fn
+
+    def get(self, spec):
+        """Some functions take a resolved spec dict as input.
```

Comment:
Can we add some comment about functions just getting resampled? No guarantee that the same thing will get restored.

Reply:
sure. let me know if this is what you had in mind.